cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,203 @@
+
+	defaults := []Provider{
+		// cluster API core provider
+		&provider{
+			name:         CAPI,
+			url:          "https://github.com/kubernetes-sigs/cluster-api/releases/latest/cluster-api-components.yaml",
+			providerType: clusterctlv1.CoreProviderType,
+		},
+
+		// Infrastructure providersClient
+		&provider{
+			name:         "aws",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "docker",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-docker/releases/latest/provider_components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "vsphere",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+
+		// Bootstrap providersClient
+		// TODO: CABPK in v1alpha3 will be included into CAPI, so this entry can be removed as soon as v1alpha3 is ready for test
+		&provider{
+			name:         "kubeadm",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/bootstrap-components.yaml",
+			providerType: clusterctlv1.BootstrapProviderType,
+		},
+	}
+
+	// ensure defaults are consistently sorted
+	sort.Slice(defaults, func(i, j int) bool {
+		return defaults[i].Name() < defaults[j].Name()
+	})
+
+	return defaults
+}
+
+func (p *providersClient) List() ([]Provider, error) {
+	// Creates a maps with all the defaults provider configurations
+	rMap := map[string]Provider{}
+	for _, r := range p.defaults() {
+		rMap[r.Name()] = r
+	}
+
+	// Gets user defined provider configurations, validate them, and merges with
+	// hard-coded configurations handling conflicts (user defined take precedence on hard-coded)
+	type configProvider struct {
+		Name string                    `json:"name,omitempty"`
+		URL  string                    `json:"url,omitempty"`
+		Type clusterctlv1.ProviderType `json:"type,omitempty"`
+	}
+
+	var ur []configProvider //nolint
+	err := p.reader.UnmarshalKey(ProvidersConfigKey, &ur)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal providers from the clusterctl configuration file")
+	}
+
+	for _, u := range ur {
+		r := NewProvider(u.Name, u.URL, u.Type)
+
+		err := validateProviderRepository(r)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error validating configuration from %q. Please fix the providers value in clusterctl configuration file", r.Name())
+		}
+
+		if _, ok := rMap[r.Name()]; ok {
+			klog.V(3).Infof("The clusterctl configuration file overrides default configuration for provider %s", r.Name())
+		}
+		rMap[r.Name()] = r
+	}
+
+	// Converts provider configurations maps to a slice
+	var providers []Provider //nolint
```

Comment:
curious why this fails linting.

Reply:
don't like the slice created without specifying a limit (BTW this bit changed)